sdks/python/apache_beam/runners/direct/evaluation_context.py

```diff
@@ -217,11 +242,13 @@ def handle_result(
             self._side_inputs_container.add_values(
                 view,
                 committed_bundle.get_elements_iterable(make_copy=True))
-          if (self.get_execution_context(result.transform)
-              .watermarks.input_watermark
-              == WatermarkManager.WATERMARK_POS_INF):
-            self._pending_unblocked_tasks.extend(
-                self._side_inputs_container.finalize_value_and_get_tasks(view))
+
+      # Tasks generated from unblocked side inputs as the watermark progresses.
+      tasks = self._watermark_manager.update_watermarks(
+          completed_bundle, result.transform, completed_timers,
+          committed_bundles, unprocessed_bundles, result.keyed_watermark_holds,
+          self._side_inputs_container)
+      self._pending_unblocked_tasks.extend(tasks)
```

Comment:
Should this stay inside the if condition? Since only when a view is updated, it will be necessary to check for unblocked tasks.

Reply:
I think we need it to be more general than only if the current bundle is a view.  Any watermark update may trigger downstream views to be updated--for example, say my windowing is Fixed(10), and I have a single element at time t=5.  When the upstream watermark hits t=10, there may not be a bundle processed by the transform directly going into the view, so we do need this in the full generality, since the watermark update of the previous step could update the watermark of the transform emitting the view.

Reply:
I pushed all the changes earlier today and was going to say the same (the update needs to be regardless). That is why this was not modified in my last commit.

Reply:
Makes sense. Thank you.